File: src/cli/commands/repo.js
import chalk from 'chalk';
import fs from 'node:fs';
import path from 'node:path';
import { findProjectRoot, loadConfig, saveConfig } from '../../lib/config.js';
import { discoverRepos } from '../../lib/discover.js';
import { getGitInfo } from '../../lib/git.js';

function getProjectContext() {
  const root = findProjectRoot();
  if (!root) {
    console.error(chalk.red('No .holoctl/ found. Run `holoctl init` first.'));
    process.exit(1);
  }
  const config = loadConfig(root);
  return { root, config };
}

export function registerRepoCommand(program) {
  const repo = program
    .command('repo')
    .alias('r')
    .description('Manage repos (sub-directories) within a project root');

  repo
    .command('add <repoPath>')
    .description('Register a sub-directory/repo in this project')
    .option('--name <name>', 'Short name for the repo (defaults to directory name)')
    .option('--description <desc>', 'Brief description', '')
    .action((repoPath, opts) => {
      const { root, config } = getProjectContext();

      const abs = path.resolve(repoPath);
      if (!fs.existsSync(abs)) {
        console.error(chalk.red(`Path does not exist: ${abs}`));
        process.exit(1);
      }

      const name = opts.name || path.basename(abs);
      const rel = path.relative(root, abs).replace(/\\/g, '/');

      if (config.project.repos.find(r => r.name === name)) {
        console.error(chalk.yellow(`Repo "${name}" already registered.`));
        process.exit(1);
      }

      config.project.repos.push({ name, path: rel, description: opts.description });
      saveConfig(root, config);

      const git = getGitInfo(abs);
      const branch = git.isGit ? chalk.cyan(git.branch) : chalk.dim('not a git repo');
      console.log(chalk.green(`Added repo "${name}"`) + `  ${chalk.dim(rel)}  ${branch}`);
    });

  repo
    .command('remove <name>')
    .description('Unregister a repo from this project')
    .action((name) => {
      const { root, config } = getProjectContext();
      const before = config.project.repos.length;
      config.project.repos = config.project.repos.filter(r => r.name !== name);
      if (config.project.repos.length === before) {
        console.error(chalk.red(`Repo "${name}" not found.`));
        process.exit(1);
      }
      saveConfig(root, config);
      console.log(chalk.green(`Removed repo "${name}"`));
    });

  repo
    .command('list')
    .alias('ls')
    .description('List repos: auto-discovered subprojects merged with manual overrides')
    .action(() => {
      const { root, config } = getProjectContext();
      const repos = discoverRepos(root, { includeManual: config.project.repos || [] });
      if (repos.length === 0) {
        console.log(chalk.dim('No subprojects discovered. Add markers (.git, package.json, …) or run `holoctl repo add <path>`.'));
        return;
      }
      for (const r of repos) {
        const abs = path.join(root, r.path);
        const git = r.git || { isGit: false };
        const exists = fs.existsSync(abs);
        const status = exists ? chalk.green('●') : chalk.red('●');
        const branch = git.isGit
          ? chalk.cyan(`[${git.branch}]`) + (git.dirty ? chalk.yellow(' *') : '')
          : chalk.dim('[no git]');
        const commit = git.commitHash ? chalk.dim(` ${git.commitHash} ${(git.lastCommit || '').slice(0, 40)}`) : '';
        const sourceTag = r.source && r.source !== 'auto' ? chalk.dim(`  (${r.source})`) : '';
        console.log(`  ${status} ${chalk.bold(r.name.padEnd(20))} ${branch}${commit}${sourceTag}`);
        console.log(`     ${chalk.dim(r.path)}`);
      }
    });

  repo
    .command('info <name>')
    .description('Show git info for a repo')
    .action((name) => {
      const { root, config } = getProjectContext();
      const entry = config.project.repos.find(r => r.name === name);
      if (!entry) {
        console.error(chalk.red(`Repo "${name}" not found.`));
        process.exit(1);
      }
      const abs = path.join(root, entry.path);
      const git = getGitInfo(abs);
      console.log(chalk.bold(`\n  ${name}\n`));
      console.log(`  Path:    ${chalk.dim(abs)}`);
      if (!git.isGit) {
        console.log(`  Git:     ${chalk.dim('not a git repository')}\n`);
        return;
      }
      console.log(`  Branch:  ${chalk.cyan(git.branch)}${git.dirty ? chalk.yellow('  (dirty)') : ''}`);
      console.log(`  Commit:  ${chalk.dim(git.commitHash)} ${git.lastCommit}`);
      console.log(`  Date:    ${chalk.dim(git.commitDate)}`);
      if (git.remote) console.log(`  Remote:  ${chalk.dim(git.remote)}`);
      console.log('');
    });
}

File: holoctl/cli/repo.py
from __future__ import annotations
from pathlib import Path
from typing import Optional

import typer
from ._console import console

from ..lib.config import find_project_root, load_config, save_config
from ..lib.discover import discover_repos
from ..lib.git import get_git_info

app = typer.Typer(help="Manage repos (sub-directories) within a project root")


def _require_ctx() -> tuple[Path, dict]:
    root = find_project_root()
    if not root:
        console.print("[red]No .holoctl/ found. Run `holoctl init` first.[/red]")
        raise typer.Exit(1)
    return root, load_config(root)


@app.command("add")
def repo_add(
    repo_path: str = typer.Argument(...),
    name: Optional[str] = typer.Option(None, "--name"),
    description: str = typer.Option("", "--description"),
):
    """Register a sub-directory/repo in this project."""
    root, config = _require_ctx()
    abs_path = Path(repo_path).resolve()

    if not abs_path.exists():
        console.print(f"[red]Path does not exist: {abs_path}[/red]")
        raise typer.Exit(1)

    repo_name = name or abs_path.name
    rel = str(abs_path.relative_to(root)).replace("\\", "/")

    if any(r["name"] == repo_name for r in config["project"].get("repos", [])):
        console.print(f"[yellow]Repo \"{repo_name}\" already registered.[/yellow]")
        raise typer.Exit(1)

    config["project"].setdefault("repos", []).append(
        {"name": repo_name, "path": rel, "description": description}
    )
    save_config(root, config)

    git = get_git_info(abs_path)
    branch = f"[cyan]{git['branch']}[/cyan]" if git["isGit"] else "[dim]not a git repo[/dim]"
    console.print(f"[green]Added repo \"{repo_name}\"[/green]  [dim]{rel}[/dim]  {branch}")


@app.command("remove")
def repo_remove(name: str = typer.Argument(...)):
    """Unregister a repo from this project."""
    root, config = _require_ctx()
    repos = config["project"].get("repos", [])
    before = len(repos)
    config["project"]["repos"] = [r for r in repos if r["name"] != name]
    if len(config["project"]["repos"]) == before:
        console.print(f"[red]Repo \"{name}\" not found.[/red]")
        raise typer.Exit(1)
    save_config(root, config)
    console.print(f"[green]Removed repo \"{name}\"[/green]")


@app.command("list")
def repo_list():
    """List repos: auto-discovered subprojects merged with manual overrides."""
    root, config = _require_ctx()
    repos = discover_repos(root, include_manual=config["project"].get("repos", []))
    if not repos:
        console.print("[dim]No subprojects discovered. Add markers (.git, package.json, …) or run `holoctl repo add <path>`.[/dim]")
        return
    for r in repos:
        abs_path = root / r["path"]
        git = r.get("git") or {"isGit": False}
        exists = abs_path.exists()
        status = "[green]●[/green]" if exists else "[red]●[/red]"
        if git.get("isGit"):
            dirty = "[yellow] *[/yellow]" if git.get("dirty") else ""
            branch = f"[cyan][{git['branch']}][/cyan]{dirty}"
            if git.get("commitHash"):
                branch += f" [dim]{git['commitHash']} {git.get('lastCommit', '')[:40]}[/dim]"
        else:
            branch = "[dim][no git][/dim]"
        source = r.get("source", "auto")
        source_tag = "" if source == "auto" else f"  [dim]({source})[/dim]"
        console.print(f"  {status} [bold]{r['name']:<20}[/bold] {branch}{source_tag}")
        console.print(f"     [dim]{r['path']}[/dim]")


@app.command("info")
def repo_info(name: str = typer.Argument(...)):
    """Show git info for a repo."""
    root, config = _require_ctx()
    entry = next((r for r in config["project"].get("repos", []) if r["name"] == name), None)
    if not entry:
        console.print(f"[red]Repo \"{name}\" not found.[/red]")
        raise typer.Exit(1)

    abs_path = root / entry["path"]
    git = get_git_info(abs_path)

    console.print(f"\n  [bold]{name}[/bold]\n")
    console.print(f"  Path:    [dim]{abs_path}[/dim]")
    if not git["isGit"]:
        console.print("  Git:     [dim]not a git repository[/dim]\n")
        return

    dirty_str = "[yellow]  (dirty)[/yellow]" if git.get("dirty") else ""
    console.print(f"  Branch:  [cyan]{git['branch']}[/cyan]{dirty_str}")
    console.print(f"  Commit:  [dim]{git['commitHash']}[/dim] {git['lastCommit']}")
    console.print(f"  Date:    [dim]{git['commitDate']}[/dim]")
    if git.get("remote"):
        console.print(f"  Remote:  [dim]{git['remote']}[/dim]")
    console.print("")
